compiler/resolution/visibleFunctions.cpp

```diff
@@ -449,131 +469,161 @@ static void lookAtTypeFirst(const char* name, CallExpr* call,
-    // a block that isn't a fn or module block.
-    visited.insert(block);
-    if (visInfo != NULL)
-      visInfo->visitedScopes.push_back(block);
+  BlockStmt* instantiationPt = getVisibleFnsInstantiationPt(block);
 
-    if (VisibleFunctionBlock* vfb = visibleFunctionMap.get(block)) {
-      // the block defines functions
+  // print tracking information
+  if (call->id == breakOnResolveID) {
+    getVisibleFnsShowBlock("methods", block, instantiationPt);
+  }
 
-      if (Vec<FnSymbol*>* fns = vfb->visibleFunctions.get(name)) {
-        forv_Vec(FnSymbol, fn, *fns) {
-          // When private methods and fields are supported, we'll need to extend
-          // this
-          visibleFns.add(fn);
-        }
-      }
-    }
+  if (firstVisit)
+    getVisibleMethodsFirstVisit(name, call, block, visInfo, visited,
+                                visibleFns);
 
-    if (block->useList != NULL) {
-      // the block uses other modules
-      for_actuals(expr, block->useList) {
-        SymExpr* se = NULL;
-        if (UseStmt* use = toUseStmt(expr)) {
-          se = toSymExpr(use->src);
-        } else if (ImportStmt* import = toImportStmt(expr)) {
-          se = toSymExpr(import->src);
-        } else {
-          INT_FATAL("bad expr in useList, expected ImportStmt or UseStmt");
-        }
-        // Intentionally ignore use/import privacy and limitations
-        // Methods, fields, and special type support functions cannot be
-        // re-enabled via only lists (or doing so is not easy for the user).
-
-        // A use statement could be of an enum instead of a module, but only
-        // modules can define functions.
-        if (ModuleSymbol* mod = toModuleSymbol(se->symbol())) {
-          if (isScopeVisibleForMethods(mod, call)) {
-            getVisibleMethodsImpl(name, call, mod->block, visInfo, visited,
-                                  visibleFns);
-          }
-        }
-      }
-    }
+  if (block->useList != NULL) {
+    getVisibleMethodsFromUseList(name, call, block, visInfo, visited,
+                                 visibleFns, inUseChain, firstVisit);
+  }
 
-    if (block->modRefs != NULL) {
-      for_actuals(expr, block->modRefs) {
-        SymExpr* se = toSymExpr(expr);
-        INT_ASSERT(se);
-        if (ModuleSymbol* mod = toModuleSymbol(se->symbol())) {
-          if (isScopeVisibleForMethods(mod, call)) {
-            getVisibleMethodsImpl(name, call, mod->block, visInfo, visited,
-                                  visibleFns);
-          }
+  if (firstVisit && block->modRefs != NULL) {
+    for_actuals(expr, block->modRefs) {
+      SymExpr* se = toSymExpr(expr);
+      INT_ASSERT(se);
+      if (ModuleSymbol* mod = toModuleSymbol(se->symbol())) {
+        if (mod->isVisible(call)) {
+          // Treat following a module reference as though it was a use statement
+          // for the purpose of determining whether we can follow private uses
+          // and imports
```

Comment:
Can you show a code sample of what this is trying to support? Is it things like `OtherModule.foo()` ?

Reply:
Yeah, that's basically it